apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,222 @@
+        private int _retries = 0;        
+
+        internal RpcRetryConsumer(IConsumer<T> consumer, RetryOptions options, 
+            Func<T, int> sizer, Func<Action, RetryOptions, ISequentialThreadingTimer> timerFactory)
+        {
+            _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _sizer = GaxPreconditions.CheckNotNull(sizer, nameof(sizer));
+            _timerFactory = GaxPreconditions.CheckNotNull(timerFactory, nameof(timerFactory));
+
+            // Start the background timer that will retry failed requests.
+            _timer = _timerFactory(FlushBuffer, _options);
+        }
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _consumer.Receive(items);
+            }
+            catch (RpcException e) when (HandleException(items, e)) { }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e) when (HandleException(items, e)) { }
+
+            return CommonUtils.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        /// <returns>True if the exception was handled, false otherwise.</returns>
+        private bool HandleException(IEnumerable<T> items, RpcException e)
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        return false;
+                    case ExceptionHandling.Ignore:
+                        Debug.WriteLine($"Ignored exception: {e}");
+                        return true;
+                    default:
+                        Debug.Fail($"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
+                        return true;
+                }
+            }
+            
+            // Get the size of items to be buffered and retried.
+            int size = items.Sum(item => _sizer(item));
+
+            // If the new items would be greater than the size of the buffer drop them or throw.
+            if (size + _sizeInBytes > _options.BufferSizeBytes) 
```

Comment:
This if block should be in the lock below. There is a race condition here where two thread can both pass this check and then both add to the size in a way that would exceed the limit.